In [1]:
import pandas as pd
import numpy as np

In [2]:
states = ['CA', 'TX', 'FL', 'NY', 'PA', 'IL', 'OH', 'GA', 'NC', 'MI', 'NJ', 'VA', 'WA', 'AZ', 'MA', 'TN']

## Dataset 5 - Plan_Attributes_PUF

In [5]:
df_plan = pd.read_csv('Data\Plan_Attributes_PUF.csv')

C:\Users\chuzd\AppData\Local\Temp\ipykernel_22788\3969581483.py:1: DtypeWarning: Columns (27,64,80,81,82,86,87,88,89,90,91,111,122,123,124,125,147,148) have mixed types. Specify dtype option on import or set low_memory=False.
  df_plan = pd.read_csv('Data\Plan_Attributes_PUF.csv')


In [7]:
df_plan = df_plan[df_plan["StateCode"].isin(states)]
df_plan.head()

BusinessYear StateCode  IssuerId      IssuerMarketPlaceMarketingName  \
697          2023        AZ     12303  DentaQuest Insurance Company, Inc.   
698          2023        AZ     12303  DentaQuest Insurance Company, Inc.   
699          2023        AZ     12303  DentaQuest Insurance Company, Inc.   
700          2023        AZ     12303  DentaQuest Insurance Company, Inc.   
701          2023        AZ     12303  DentaQuest Insurance Company, Inc.   

    SourceName      ImportDate MarketCoverage DentalOnlyPlan  \
697       HIOS  6/14/2022 1:01     Individual            Yes   
698       HIOS  6/14/2022 1:01     Individual            Yes   
699       HIOS  6/14/2022 1:01     Individual            Yes   
700       HIOS  6/14/2022 1:01     Individual            Yes   
701       HIOS  6/14/2022 1:01     Individual            Yes   

    StandardComponentId               PlanMarketingName  ...  \
697      12303AZ0010001  DentaQuest PPO  Pediatric High  ...   
698      12303AZ0010001  DentaQuest PPO  Pediatric High  ...   
699      12303AZ0010003      DentaQuest PPO Family High  ...   
700      12303AZ0010003      DentaQuest PPO Family High  ...   
701      12303AZ0010004       DentaQuest PPO Family Low  ...   

    TEHBDedOutOfNetFamilyPerPerson TEHBDedOutOfNetFamilyPerGroup  \
697                            NaN                           NaN   
698                            NaN                           NaN   
699                            NaN                           NaN   
700                            NaN                           NaN   
701                            NaN                           NaN   

    TEHBDedCombInnOonIndividual TEHBDedCombInnOonFamilyPerPerson  \
697                         NaN                              NaN   
698                         NaN                              NaN   
699                         NaN                              NaN   
700                         NaN                              NaN   
701                         NaN                              NaN   

    TEHBDedCombInnOonFamilyPerGroup IsHSAEligible  \
697                             NaN           NaN   
698                             NaN           NaN   
699                             NaN           NaN   
700                             NaN           NaN   
701                             NaN           NaN   

    HSAOrHRAEmployerContribution HSAOrHRAEmployerContributionAmount  \
697                          NaN                                NaN   
698                          NaN                                NaN   
699                          NaN                                NaN   
700                          NaN                                NaN   
701                          NaN                                NaN   

    URLForSummaryofBenefitsCoverage  \
697                                   
698                                   
699                                   
700                                   
701                                   

                                          PlanBrochure  
697  https://dentaquest.com/DentaQuest/media/PDF/Ce...  
698  https://dentaquest.com/DentaQuest/media/PDF/Ce...  
699  https://dentaquest.com/DentaQuest/media/PDF/Ce...  
700  https://dentaquest.com/DentaQuest/media/PDF/Ce...  
701  https://dentaquest.com/DentaQuest/media/PDF/Ce...  

[5 rows x 151 columns]

In [31]:
df_plan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19853 entries, 697 to 29454
Columns: 151 entries, BusinessYear to PlanBrochure
dtypes: float64(6), int64(5), object(140)
memory usage: 23.0+ MB


In [29]:
pd.DataFrame(df_plan.isna().sum()).head(60)

0
BusinessYear                                            0
StateCode                                               0
IssuerId                                                0
IssuerMarketPlaceMarketingName                          0
SourceName                                              0
ImportDate                                              0
MarketCoverage                                          0
DentalOnlyPlan                                          0
StandardComponentId                                     0
PlanMarketingName                                       0
HIOSProductId                                           0
NetworkId                                               0
ServiceAreaId                                           0
FormularyId                                           793
IsNewPlan                                               0
PlanType                                                0
MetalLevel                                              0
DesignType                                            939
UniquePlanDesign                                      793
QHPNonQHPTypeId                                         0
IsNoticeRequiredForPregnancy                          793
IsReferralRequiredForSpecialist                       783
SpecialistRequiringReferral                         11934
PlanLevelExclusions                                 17574
IndianPlanVariationEstimatedAdvancedPaymentAmou...    937
CompositeRatingOffered                                  0
ChildOnlyOffering                                       0
ChildOnlyPlanId                                     19851
WellnessProgramOffered                                791
DiseaseManagementProgramsOffered                     3329
EHBPercentTotalPremium                                941
EHBPediatricDentalApportionmentQuantity             19240
IsGuaranteedRate                                    19060
PlanEffectiveDate                                       0
PlanExpirationDate                                   9607
OutOfCountryCoverage                                    0
OutOfCountryCoverageDescription                      5266
OutOfServiceAreaCoverage                                0
OutOfServiceAreaCoverageDescription                  3363
NationalNetwork                                         0
URLForEnrollmentPayment                                 0
FormularyURL                                          793
PlanId                                                  0
PlanVariantMarketingName                                0
CSRVariationType                                        0
IssuerActuarialValue                                13069
AVCalculatorOutputNumber                             3022
MedicalDrugDeductiblesIntegrated                      793
MedicalDrugMaximumOutofPocketIntegrated               793
MultipleInNetworkTiers                                  0
FirstTierUtilization                                    0
SecondTierUtilization                               14589
SBCHavingaBabyDeductible                              793
SBCHavingaBabyCopayment                               793
SBCHavingaBabyCoinsurance                             793
SBCHavingaBabyLimit                                   793
SBCHavingDiabetesDeductible                           793
SBCHavingDiabetesCopayment                            793
SBCHavingDiabetesCoinsurance                          793
SBCHavingDiabetesLimit                                793

In [9]:
print("Filas: ", df_plan.shape[0])
print("Columnas: ", df_plan.shape[1])

Filas:  19853
Columnas:  151


Contiene 151 columnas.

In [8]:
df_plan.StateCode.unique()

array(['AZ', 'FL', 'GA', 'IL', 'MI', 'NC', 'OH', 'TN', 'TX', 'VA'],
      dtype=object)


## Plan Atributes


El Plan-PUF contiene datos de nivel de variante del plan sobre pagos máximos de bolsillo, deducibles, elegibilidad de cuenta de ahorros para la salud (HSA) y otros atributos del plan

- Busines Year: año en que se provee el seguro*
- StateCode: estado donde se está ofertado 
- IssurId: identificador de emisora de seguro HIOS 
- IssuerMarketplaceMarketingName: nombre comercial de la aseguradora
- SourceName: fuente de la data*
- ImportDate: eliminar*
- DentalOnlyPlan: indica si el plan solo es dental (puede servir para eliminar los planes dentales)
- StandardComponentId: indenficador del plan.
- PlanMarketingName: nombre del plan de seguro 
- HIOSProductId: numero que indetifica el producto del seguro (el producto cualquier compromiso de la aseguradora)
- NetworkId: indenfica de la red de cuidados del seguro.
- ServiceAreaId: area de servicio
- FormularyId: id de formulario de medicamentos
- IsNewPlan: indica si el seguro es nuevo en el año actual (bool)
- PlanType: tipo de plan (varios valores)
- MetalLevel: es la categoría de covertura del seguro
- DesignType: si el plan es estandarizado para su MetalLevel
- UniquePlanDesign: si el plan tien un unico diseño (bool)
- IssuerActuarialValue: porcentaje de lo que cubre el seguro.
- 



Se identifica la categoría de: 
indentificadores, 
categoricas sí o no, 
categoricas de tipos
Por diabetes
por familia
por individuo




In [33]:
df_plan.IssuerMarketPlaceMarketingName.nunique()

121

Significa que solo hay 121 aseguradoras en la data de los estados que observamos.

In [34]:
df_plan.SourceName.unique()

array(['HIOS', 'SERFF'], dtype=object)

In [42]:
df_plan["DentalOnlyPlan"].unique()

array(['Yes', 'No'], dtype=object)

In [45]:
df_plan["PlanMarketingName"].nunique()

1909

In [48]:
df_plan["PlanType"].unique()

array(['PPO', 'HMO', 'EPO', 'POS', 'Indemnity'], dtype=object)

In [49]:
df_plan["DesignType"].unique()

array([nan, 'Not Applicable', 'Design 1'], dtype=object)

In [50]:
df_plan["UniquePlanDesign"].unique()

array([nan, 'Yes', 'No'], dtype=object)

In [52]:
df_plan["PlanId"].nunique()

# idenficador de los planes, este es imporante también

19853

In [54]:
df_plan["IssuerActuarialValue"].unique()


array([nan, '62.98%', '100.00%', '62.76%', '64.98%', '94.04%', '87.23%',
       '70.31%', '73.30%', '94.06%', '73.38%', '87.17%', '70.38%',
       '64.84%', '78.23%', '73.33%', '94.11%', '71.27%', '87.27%',
       '64.73%', '94.08%', '73.39%', '70.33%', '70.29%', '94.22%',
       '87.19%', '73.14%', '71.21%', '73.40%', '94.09%', '78.45%',
       '81.32%', '73.86%', '70.23%', '94.05%', '87.55%', '64.45%',
       '60.74%', '73.48%', '94.02%', '70.49%', '78.18%', '70.00%',
       '64.14%', '80.38%', '70.13%', '87.36%', '94.58%', '73.90%',
       '70.47%', '87.40%', '73.45%', '94.42%', '71.35%', '73.37%',
       '87.76%', '94.96%', '78.07%', '70.03%', '73.03%', '94.03%',
       '87.04%', '64.85%', '64.88%', '73.02%', '94.10%', '70.06%',
       '87.01%', '63.25%', '61.05%', '78.59%', '71.22%', '94.13%',
       '87.05%', '78.75%', '70.01%', '73.04%', '94.07%', '70.07%',
       '87.24%', '71.62%', '73.83%', '70.57%', '95.00%', '73.53%',
       '87.93%', '73.65%', '70.56%', '87.82%', '94.93%',

In [57]:
df_plan["AVCalculatorOutputNumber"].unique()


array([       nan, 0.62714523, 1.        , ..., 0.70007372, 0.94116285,
       0.73598641])

## Dateset 6 - Rata_PUF

## Rate_PUF

Esta tabla contiene la tarifas de los seguros con base en su fumador, la ubicacion geográfica del asegurado y tarifas a nivel familiar. 

- BusinessYear: elminar*
- StateCode. codigo de estado 
- IssuerId: id de aseguradora 
- SourceName: fuente de datos* 
- ImportDate: fecha de impostacion*
- RateEffectiveDate: fecha de cuando el seguro se hace efectivo*
- RateExpirationDate: fecha de espiración del seguro*
- PlanId: id del plan de seguro. 
- RatingAreaId: areas geográficas dentro del estado, para calificacion.
- Tobacco: si es un fumador*
- Age: rango de edad del suscriptor del seguro o dependientes
- IndividualRate: cobro a individual 
- IndividualTobaccoRate: tarifa a individual que consume tabaco
- Couple: suscriptor y pareja
- PrimarySubscriberAndOneDependent: suscriptor y un dependediente 
- PrimarySubscriberAndTwoDependents: suscriptor y dos dependientes 
- PrimarySubscriberAndThreeOrMoreDependents: suscriptor y tres o más dependientes
- CoupleAndOneDependent: suscriptor, pareja y un dependiente 
- CoupleAndTwoDependents: suscriptor, pareja y dos dependientes
- CoupleAndThreeOrMoreDependent: suscriptor, pareja y tres o más dependientes


Las columnas de IndividualRate en adelante representa la tarifa que se cobra por esa variación del seguro, podemos crear una nueva variable que sea tipo de seguro. Por ejemplo: 

- individual
- individual tabbaco
- Paraje, un dependiente
- dos dependientes
- tres o más dependientes
- pareja y dependetes, etc. 


In [3]:
df_rate = pd.read_csv('Data\Rate_PUF.csv')

In [4]:
df_rate = df_rate[df_rate["StateCode"].isin(states)]
df_rate.head()

BusinessYear StateCode  IssuerId SourceName           ImportDate  \
64039          2023        AZ     12303       HIOS  2022-06-14 01:01:37   
64040          2023        AZ     12303       HIOS  2022-06-14 01:01:37   
64041          2023        AZ     12303       HIOS  2022-06-14 01:01:37   
64042          2023        AZ     12303       HIOS  2022-06-14 01:01:37   
64043          2023        AZ     12303       HIOS  2022-06-14 01:01:37   

      RateEffectiveDate RateExpirationDate          PlanId   RatingAreaId  \
64039        2023-01-01         2023-12-31  12303AZ0010001  Rating Area 1   
64040        2023-01-01         2023-12-31  12303AZ0010001  Rating Area 2   
64041        2023-01-01         2023-12-31  12303AZ0010001  Rating Area 3   
64042        2023-01-01         2023-12-31  12303AZ0010001  Rating Area 4   
64043        2023-01-01         2023-12-31  12303AZ0010001  Rating Area 5   

             Tobacco   Age  IndividualRate  IndividualTobaccoRate  Couple  \
64039  No Preference  0-14           29.78                    NaN     NaN   
64040  No Preference  0-14           29.78                    NaN     NaN   
64041  No Preference  0-14           29.78                    NaN     NaN   
64042  No Preference  0-14           29.78                    NaN     NaN   
64043  No Preference  0-14           29.78                    NaN     NaN   

       PrimarySubscriberAndOneDependent  PrimarySubscriberAndTwoDependents  \
64039                               NaN                                NaN   
64040                               NaN                                NaN   
64041                               NaN                                NaN   
64042                               NaN                                NaN   
64043                               NaN                                NaN   

       PrimarySubscriberAndThreeOrMoreDependents  CoupleAndOneDependent  \
64039                                        NaN                    NaN   
64040                                        NaN                    NaN   
64041                                        NaN                    NaN   
64042                                        NaN                    NaN   
64043                                        NaN                    NaN   

       CoupleAndTwoDependents  CoupleAndThreeOrMoreDependents  
64039                     NaN                             NaN  
64040                     NaN                             NaN  
64041                     NaN                             NaN  
64042                     NaN                             NaN  
64043                     NaN                             NaN

In [5]:
print("Filas: ", df_rate.shape[0])
print("Columnas: ", df_rate.shape[1])

Filas:  1729685
Columnas:  20


In [16]:
df_rate.isna().sum()

BusinessYear                                       0
StateCode                                          0
IssuerId                                           0
SourceName                                         0
ImportDate                                         0
RateEffectiveDate                                  0
RateExpirationDate                                 0
PlanId                                             0
RatingAreaId                                       0
Tobacco                                          683
Age                                                0
IndividualRate                                     0
IndividualTobaccoRate                         812909
Couple                                       1729002
PrimarySubscriberAndOneDependent             1729002
PrimarySubscriberAndTwoDependents            1729002
PrimarySubscriberAndThreeOrMoreDependents    1729002
CoupleAndOneDependent                        1729002
CoupleAndTwoDependents                       1

In [7]:
df_rate["RatingAreaId"].unique()

array(['Rating Area 1', 'Rating Area 2', 'Rating Area 3', 'Rating Area 4',
       'Rating Area 5', 'Rating Area 6', 'Rating Area 7', 'Rating Area 8',
       'Rating Area 9', 'Rating Area 10', 'Rating Area 11',
       'Rating Area 12', 'Rating Area 13', 'Rating Area 14',
       'Rating Area 15', 'Rating Area 16', 'Rating Area 17',
       'Rating Area 18', 'Rating Area 19', 'Rating Area 20',
       'Rating Area 21', 'Rating Area 22', 'Rating Area 23',
       'Rating Area 24', 'Rating Area 25', 'Rating Area 26',
       'Rating Area 27', 'Rating Area 28', 'Rating Area 29',
       'Rating Area 30', 'Rating Area 31', 'Rating Area 32',
       'Rating Area 33', 'Rating Area 34', 'Rating Area 35',
       'Rating Area 36', 'Rating Area 37', 'Rating Area 38',
       'Rating Area 39', 'Rating Area 40', 'Rating Area 41',
       'Rating Area 42', 'Rating Area 43', 'Rating Area 45',
       'Rating Area 46', 'Rating Area 47', 'Rating Area 48',
       'Rating Area 49', 'Rating Area 50', 'Rating Area 51

In [8]:
df_rate["Age"].unique()

array(['0-14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24',
       '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35',
       '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46',
       '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57',
       '58', '59', '60', '61', '62', '63', '64 and over', 'Family Option'],
      dtype=object)

In [11]:
df_rate["IndividualRate"].unique()

array([ 29.78,   0.  ,  20.8 , ..., 210.08, 213.26,  48.44])

## Dataset 7 - Service_Area_PUF

## Service Area

- BusinessYear: listo
- StateCode: listo
- IssuerId: listo
- SourceName: listo
- ImportDate: listo
- ServiceAreaId: 
- ServiceAreaName: 
- CoverEntireState: 
- County: 
- PartialCounty: 
- ZipCodes: 
- PartialCountyJustification: 
- MarketCoverage: 
- DentalOnlyPla: 

In [17]:
df_service = pd.read_csv("Data/Service_Area_PUF.csv")

In [18]:
df_service.head()

BusinessYear StateCode  IssuerId SourceName           ImportDate  \
0          2023        AK     21989       HIOS  2022-11-19 01:01:13   
1          2023        AK     21989       HIOS  2022-11-19 01:01:13   
2          2023        AK     21989       HIOS  2022-11-19 01:01:13   
3          2023        AK     21989       HIOS  2022-11-19 01:01:13   
4          2023        AK     21989       HIOS  2022-11-19 01:01:13   

  ServiceAreaId ServiceAreaName CoverEntireState  County PartialCounty  \
0        AKS001  Alaska Premier              Yes     NaN           NaN   
1        AKS001  Alaska Premier              Yes     NaN           NaN   
2        AKS002      Alaska PPO               No  2020.0            No   
3        AKS002      Alaska PPO               No  2020.0            No   
4        AKS002      Alaska PPO               No  2090.0            No   

  ZipCodes PartialCountyJustification      MarketCoverage DentalOnlyPlan  
0      NaN                        NaN          Individual            Yes  
1      NaN                        NaN  SHOP (Small Group)            Yes  
2      NaN                        NaN          Individual            Yes  
3      NaN                        NaN  SHOP (Small Group)            Yes  
4      NaN                        NaN  SHOP (Small Group)            Yes

In [20]:
df_service = df_service[df_service["StateCode"].isin(states)]
df_service.head()

BusinessYear StateCode  IssuerId SourceName           ImportDate  \
198          2023        AZ     12303       HIOS  2022-06-14 01:01:37   
199          2023        AZ     13877       HIOS  2023-01-26 01:01:51   
200          2023        AZ     13877       HIOS  2023-01-26 01:01:51   
201          2023        AZ     17100       HIOS  2022-06-10 01:01:25   
202          2023        AZ     17454       HIOS  2022-07-16 01:01:01   

    ServiceAreaId                          ServiceAreaName CoverEntireState  \
198        AZS001                          AZ Entire State              Yes   
199        AZS001                            Oscar Arizona               No   
200        AZS001                            Oscar Arizona               No   
201        AZS001         METLIFE PREFERRED DENTAL PROGRAM              Yes   
202        AZS001  GLIC Individual PPO On and Off Exchange              Yes   

     County PartialCounty ZipCodes PartialCountyJustification  \
198     NaN           NaN      NaN                        NaN   
199  4013.0            No      NaN                        NaN   
200  4021.0            No      NaN                        NaN   
201     NaN           NaN      NaN                        NaN   
202     NaN           NaN      NaN                        NaN   

         MarketCoverage DentalOnlyPlan  
198          Individual            Yes  
199          Individual             No  
200          Individual             No  
201  SHOP (Small Group)            Yes  
202          Individual            Yes